### PR TITLE
Feature/healthcheck

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

- Converted to Go module
- Updated CI to use go 1.13.8, and updated /health endpoint in nomad plan
- Added new healthcheck library
- Makefile now generates always the same file to prevent duplicates, and it is ignored by git (will be generated as part of the build)
- Implemented graceful shutdown with timeout

### How to review

- Make sure code changes make sense
- Make sure unit tests pass
- You can try the /health endpoint (`GET http://localhost:20010/health`)
- Check that graceful shutdown works - you can optionally add a time.Sleep(*time.Second) in the shutdown goroutine, to trigger the timeout.
- In an shell without having exported GOPATH, try to run `make build` and `make run`. Validate that a file assets/data.go is generated, with encoded data in the first case and links to asset files in the second case.

### Who can review

Anyone.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,6 +11,9 @@ dp-frontend-renderer
 | ENABLE_COOKIES_CONTROL        | false                 | Enabled user cookie preferences
 | PATTERN_LIBRARY_ASSETS_PATH   |                       | Pattern library location
 | SUPPORTED_LANGUAGES           | [2]string{"en", "cy"} | languages supported
+| GRACEFUL_SHUTDOWN_TIMEOUT     | 5s                    | The graceful shutdown timeout
+|	HEALTHCHECK_INTERVAL          | 10s                   | Time between calls to healthchecks
+|	HEALTHCHECK_RECOVERY_INTERVAL | 1m                    | Timeout to consider a failing healthcheck critical
```
Comment (CarlHembrough) [inline README.md:16]: should be `HEALTHCHECK_CRITICAL_TIMEOUT`? 

(also replace the tab at the start with a space? - not sure if this makes a difference when rendered)

File: main.go [hunk 2]
```diff
@@ -31,9 +47,59 @@ func init() {
...
+		cancel()
+	}()
+
+	// wait for Shutdown timeout or success (via cancel)
+	<-shutdownCtx.Done()
+	if shutdownCtx.Err() == context.DeadlineExceeded {
+		log.Event(shutdownCtx, "shutdown timeout", log.ERROR, log.Error(shutdownCtx.Err()))
 		os.Exit(1)
 	}
+	log.Event(shutdownCtx, "done shutdown gracefully", log.ERROR, log.Data{"context": shutdownCtx.Err()})
```
Comment (CarlHembrough) [inline main.go:103]: should be log.Info?

File: service/service.go [hunk 3]
```diff
@@ -59,24 +49,37 @@ func Run(ctx context.Context, taxonomyRedirects map[string]string) error {
...
 		WriteTimeout: 10 * time.Second,
 	}
 
-	if err = server.ListenAndServe(); err != nil {
-		return errors.Wrap(err, "failed to start server")
+	// Start the HTTP server in a new go routine
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			log.Event(ctx, "error starting http server", log.FATAL, log.Error(err))
+			os.Exit(1)
```
Comment (CarlHembrough) [inline service/service.go:73]: Should we exit here? In other services I think we just logged a message?

Comment (DavidSubiros) [reply service/service.go:73]: yes, thanks!

File: LICENSE.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 The MIT License (MIT)
 
-Copyright (c) 2016 Office for National Statistics
+Copyright (c) 2020 Office for National Statistics
```
Comment (nshumoogum) [inline LICENSE.md:3]: Probably should be changed to `2016-2020`

File: Makefile [hunk 5]
```diff
@@ -1,24 +1,31 @@
...
+LDFLAGS=-ldflags "-w -s -X 'main.Version=${VERSION}' -X 'main.BuildTime=$(BUILD_TIME)' -X 'main.GitCommit=$(GIT_COMMIT)'"
+
+build: generate-prod
+	go build $(LDFLAGS) -tags 'production' -o $(BINPATH)/dp-frontend-renderer
 	cp taxonomy-redirects.yml $(BINPATH)
 
-debug: generate
-	go build -tags 'debug' -o build/dp-frontend-renderer
+debug: generate-debug
+	go build $(LDFLAGS) -tags 'debug' -o build/dp-frontend-renderer
```
Comment (nshumoogum) [inline Makefile:13]: I would add the `-race` flag into debug command

File: README.md [hunk 6]
```diff
@@ -11,6 +11,9 @@ dp-frontend-renderer
 | ENABLE_COOKIES_CONTROL        | false                 | Enabled user cookie preferences
 | PATTERN_LIBRARY_ASSETS_PATH   |                       | Pattern library location
 | SUPPORTED_LANGUAGES           | [2]string{"en", "cy"} | languages supported
+| GRACEFUL_SHUTDOWN_TIMEOUT     | 5s                    | The graceful shutdown timeout
+|	HEALTHCHECK_INTERVAL          | 30s                   | Time between calls to healthchecks
```
Comment (nshumoogum) [inline README.md:15]: Some alignment issues

File: go.sum [hunk 7]
```diff
@@ -0,0 +1,103 @@
+github.com/BurntSushi/toml v0.3.0/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
```
Comment (nshumoogum) [inline go.sum:1]: You may have done this already but just run `go mod tidy`

File: main.go [hunk 8]
```diff
@@ -31,9 +47,59 @@ func init() {
 }
 
 func main() {
+	log.Namespace = namespace
 	ctx := context.Background()
-	if err := service.Run(ctx, taxonomyRedirects); err != nil {
```
Comment (nshumoogum) [inline main.go:35]: So we should move the code from main to exist in service.Run(...) function, see my [dataset API pr](https://github.com/ONSdigital/dp-dataset-api/pull/215)

File: main.go [hunk 9]
```diff
@@ -31,9 +47,59 @@ func init() {
...
+	// Get Config
+	cfg, err := config.Get()
+	if err != nil {
+		log.Event(ctx, "unable to retrieve service configuration", log.FATAL, log.Error(err))
+		os.Exit(1)
+	}
+	log.Event(ctx, "got service configuration", log.Data{"config": cfg}, log.INFO)
+
+	// Make os Signal channels
+	signals := make(chan os.Signal, 1)
```
Comment (nshumoogum) [inline main.go:62]: This should be setup before `config.Get()`

File: main.go [hunk 10]
```diff
@@ -31,9 +47,59 @@ func init() {
...
+		log.Event(shutdownCtx, "stopping healthcheck", log.INFO)
+		hc.Stop()
+
+		log.Event(shutdownCtx, "stopping http server", log.INFO)
+		if err := httpServer.Shutdown(shutdownCtx); err != nil {
+			log.Event(shutdownCtx, "failed to gracefully shutdown http server", log.FATAL, log.Error(err))
+			os.Exit(1)
+		}
+
+		cancel()
```
Comment (nshumoogum) [inline main.go:94]: I would use `defer cancel()` at line 85

File: main.go [hunk 11]
```diff
@@ -31,9 +47,59 @@ func init() {
...
+
+	// Shutdown goroutine
+	go func() {
+		log.Event(shutdownCtx, "stopping healthcheck", log.INFO)
+		hc.Stop()
+
+		log.Event(shutdownCtx, "stopping http server", log.INFO)
+		if err := httpServer.Shutdown(shutdownCtx); err != nil {
+			log.Event(shutdownCtx, "failed to gracefully shutdown http server", log.FATAL, log.Error(err))
+			os.Exit(1)
```
Comment (nshumoogum) [inline main.go:91]: I wouldn't exit here but capture that the graceful shutdown has failed, again see [dataset API pr](https://github.com/ONSdigital/dp-dataset-api/pull/215)

File: README.md (hunk 6 above)
Comment (DavidSubiros) [reply README.md:15]: it's strange - I don't see it in my vscode:
image

Comment (nshumoogum) [reply README.md:15]: If you delete the whitespace inbetween | and Health<...> and then add space does git diff notice the difference?

File: go.sum (hunk 7 above)
Comment (DavidSubiros) [reply go.sum:1]: yes - but then it removes the dependency with go-bindata, which is actually needed. I imagine it has something to do with the fact that it go-bindata doesn't use go modules. I checked that doing a go mod tidy, it only removes this dependency (there are no more unused dependencies in go.sum)

**Pull request merged**